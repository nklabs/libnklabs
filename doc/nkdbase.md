# Database based on serialization library

Functions are provided to create a database located in flash memory.  This
just means that we save a serialized form of a structure into flash memory. 
The serialized form is useful because it tends to preserve data across
schema changes.  See [nkserialize](nkserialize.md) for the serialization
format.  [nkchecked](nkchecked.md) is used to wrap the serialized data with
a header containing size and CRC in flash memory.

Two areas of memory (banks) are allocated so that an older version of the
database is always preserved in case of a power outage during a write.

Each bank contains an 8 byte header (from [nkchecked](nkchecked.md))
continaing the file size and CRC followed by a one byte revision code,
followed by the serialized database string (NUL terminated).

## Files

[nkdbase.h](../inc/nkdbase.h), [nkdbase.c](../src/nkdbase.c)

## Usage

An nk_dbase structure should be filled out with information about the database:

~~~c
	const struct nk_dbase test_dbase =
	{
	    .ty = &tyTESTTOP,			// Schema for database
	    .bank0 = 65536+32768+0,		// Location of bank0 in flash memory
	    .bank1 = 65536+32768+8192,		// Location of bank1 in flash memory
	    .bank_size = 8192,			// Size of each flash memory bank (this size is used for flash_erase)
	    .buf = xfer_buf,			// Transfer buffer: used for flash_read and flash_write
	    .buf_size = sizeof(xfer_buf),	// Sizeof transfer buffer
	    .flash_read = nk_mcuflash_read,	// Flash read access function
	    .flash_erase = nk_mcuflash_erase,	// Flash erase function
	    .flash_write = nk_mcuflash_write,	// Flash write function
	    .flash_granularity = 8		// Write granularity- writes are padded so that they always a multiple of this size
						// 1 is allowed for granularity
	};
~~~

## nk_dbase_load()

~~~c
int nk_dbase_load(
	const struct nk_dbase *dbase,
	char *rev, // Address of version number
	void *ram // Address of deserialized database in RAM
);
~~~

Load a database into RAM.  The revision number of the loaded database is
saved in *rev.

Returns zero for success.

__nk_fscan__ (see [nkserialize](nkserialize.md)) is used to deserialize the
database into a buffer (address given by the 'ram' parameter) according to
the schema in nk_dbase.ty.

## nk_dbase_save()

~~~c
int nk_dbase_save(
	const struct nk_dbase *dbase,
	char *rev, // Address of version number
	void *ram // Address of deserialized database in RAM
);
~~~

Save a database from RAM into flash.  This bumps the revision number if the
database was successfully saved.

Returns zero for success.

__nk_dbase_serialize__ (see [nkserialize](nkserialize.md)) is used to
serialize the database contained in a buffer (address in 'ram' parameter)
into a string for the flash memory.  nk_dbase.ty hold the schema.

## Configuration or calibration database

A template CLI command along with example database schema is provided which
shows how a user interface for a database might work, see
[../app/database.c](../app/database.c)

The dbase command allows you to query or set elements of the database.  The
advantage of this method is that it avoids the need for writing a user
interface for each configuration variable.

A typical CLI command looks like this:

	dbase show			Print entire database in serialized format
	dbase replace			Replace the entire database with following multi-line input
	dbase get a.b			Print a particular element
	dbase set a.b 7		        Set a particular element
	dbase set a { b:7, c:6 }	Set an entire sub-structure
	dbase set a			Set an entire sub-structure with following multi-line input
	dbase save			Save database to flash
	dbase load			Load database from flash
	dbase reset			Reset database to default values
