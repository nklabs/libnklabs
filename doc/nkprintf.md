# Formatted output

## Files

[nkprintf.h](../inc/nkprintf.h),
[nkprintf.c](../src/nkprintf.c),
[nkprintf_fp.c](../src/nkprintf_fp.c)

## Description

```c
int nk_printf(const char *fmt,...);

int nk_fprintf(nkoutfile_t *f, const char *fmt, ...);

int nk_snprintf(char *dest, size_t len, const char *fmt,...);
int nk_sfprintf(char *dest, size_t len, const char *fmt,...);

int nk_indent(int ind);

int nk_findent(nkoutfile_t *f, int ind);

int nk_byte_hex_dump(unsigned char *buf, unsiged long base, unsigned long offset, unsigned long count);

int nk_fbyte_hex_dump(nkoutfile_t *f, unsigned char *buf, unsigned long base, unsigned long offset, unsigned long count);

```

These functions provide low memory overhead formatted printing.  They
provide a subset of the C-standard printf, but use less memory space. 
Floating point is optionally supported, and even when enabled, much less
space will be used than newlib printf.

__nk_printf__ prints to __nkstdout__ using __nk_fputc__.

__nk_fprintf__ prints to the __nkoutfile_t__ specified in the __f__
argument using __nk_fputc__.

__nk_printf__ and __nk_fprintf__ return the bitwise-OR of the return values
of all calls made to __nk_fputc__.  A return value of 0 means no error.

+__nk_snprintf__ prints a C-string to a memory buffer.  The terminating NUL
is always included.  If the memory area overflows, __nk_snprintf__ returns
true (otherwise false is returned).  Even if the buffer overflows, the
terminating NUL is written.

__nk_sfprintf__ prints a space-filled field to a memory buffer.  It prints
to the specified memory buffer, and then the balance of the buffer is filled
with spaces.  If the memory area overflows, __nk_sfprintf__ returns true,
otherwise false is returned.

__nk_indent__ prints __ind__ spaces to the __nkstdout__ using __nk_fputc__.

__nk_findent__ prints __ind__ spaces to the specified nkoutfile_t using
__nk_fputc__.

__nk_indent__ and __nkfindent__ return the bitwise-OR of the return values
of all calls made to __nk_fputc__.  A return value of zero means no error.

__nk_byte_hex_dump__ prints a hex/ASCII dump to __nkstdout__ using
__nk_fputc__.  __nk_fbyte_hex_dump__ is the same, except the output is to
the specified __nkoutfile_t__.

__count__ is the number of bytes to dump.

__offset__ is the starting byte offset within the __buf__ to dump.

__base__ is added to __offset__ for the printed address.

__nk_byte_hex_dump__ and __nk_fbyte_hex_dump__ return the bit-wise OR of the
return values of all calls to __nk_fputc__.  A return value of 0 means no
error.

## Format string reference

All characters are printed as-is, except for '%' which introduces a format
specifier.  Use '%%' to print '%'.  The last character of the format
specifier should be one of these:

* c prints a character
* s prints a C string
* d prints an int in decimal
* u prints an unsigned int in decimal
* x prints an unsigned int in lower case hexadecimal
* X print an unsigned int in upper case hexadecimal
* p prints a pointer
* f prints a double precision floating point number: 0.000120
* e, E prints a double precision floating point number: 1.200000e-4, 1.200000E-4
* g, G prints a double precision floating point number: 0.0012

Flag characters may optionally appear after the %:

* '-' Left justifies text within field, otherwise right justify
* '+' A plus sign or minus sign prefixes the number
* ' ' A space prefixes a positive number
* '#' Alternate form (hexadecimal prefixed with 0x).
* '0' Zero pad to the left

Next, a field width may optionally appear (NNN).  '*' may be given instead-
it means the field width is given as the integrer argument.

Next, a period followed by a precision may optionally appear (.NNN).  '.*'
may be given instead- it means the precision is given as an integer
argument.

For integer formats, an int or unsigned int argument is assumed, but this
can be overridden:

* hh for a char
* h for a short
* l for a long
* ll for a long long
 