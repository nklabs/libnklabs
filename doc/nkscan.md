# Command Line Argument Parser

## Files

[nkscan.h](../inc/nkscan.h),
[nkscan.c](../src/nkscan.c),
[nkstrtod.c](../src/nkstrtod.c)

## Description

## nk_scan()

```c
int nk_fscan(nkinfile_t *f, const char *fmt, ...);
int nk_fiscan(nkinfile_t *f, const char *fmt, ...);

```

These are enhanced versions of C's scanf().  The main enhancement is that
they either fully parse the input as specified by the format string and
return true, or return false but leave the input nkinfile_t left at its
original position.  This allows you to make multiple __nk_fscan__
calls until a correct match is found.  These were designed as a convenient way to
parse command line arguments.

Since these functions may reposition the nkinfile_t, the nkinfile_t must
support random access.

__nk_fscan__ is case sensitive.

__nk_fiscan__ is case insensitive and unquoted words and identifiers (%w and
%i) are quashed to lower case.

## Format string

Each character in the format string must match the next character from the input string, except:

Space matches zero or more spaces or tabs

%%  Matches %

%hhd   Matches a decimal number, stores result into a char.  A (char *) should be provided.

%hd    Matches a decimal number, sotres result into a short.  A (short *) should be provided.

%d     Matches a decimal number, stores result into an int.  An (int *) should be provided.

%ld    Matches a decimal number, stores result into a long.  A (long *) should be provided.

%lld   Matches a decimal number, stores result into a long long.  A (long long *) should be provided.

%zd    Matches a decimal number, stores result into a size_t.  A (size_t *) should be provided.

%td    Matches a decimal number, stores result into a ptrdiff_t.  A (ptrdiff_t *) should be provided.

%hhx etc.   Same as above, but hexadecimal number into unsigned

%hhu etc.   Same as above, but decimal number into unsigned

%f     Matches a floating point number.  A (double *) should be provided.

%p     Matches a pointer in the form of a hexadecimal number, stores result into a void *.  A (void **) should be provided.

%i     Save identifier in a string buffer.  The identifier is composed of [a-zA-Z_][a-zA-Z_0-9]*  Both the address of the string buffer (char *) and the length of the string buffer (size_t) should be provided.

If the identifier begins with ", then assume it's a C string.  All
characters up to the following " are included in the identifier, and C
string escape sequences are available from within the string.

%w     Save word in a string buffer.  The word is composed of non-whitespace chars followed by whitespace or end of input.  Both the address of the string (char *) and the length of the string (size_t) should be provided.

If the word begins with ", then assume it's a C string.  All characters up
to the following " are included in the word, and C string escape sequences
are available from within the string.

%e     Normally the end of the format string must correspond to the end of the input (otherwise it will fail to parse). 
But if the last thing in the format string is %e, then parse will succeed
even if there is more input available.  This allows you to parse more of the
input with another call to nk_fscan.

A field width may be specified after any %, like %2d.  This will limit the
parsing to the indicated size.  WIth %2d, only the first two digits will be
taken.  For example, you could parse a date such as 20181231 with
"%4d%2d%2d", &year, &month, &day
