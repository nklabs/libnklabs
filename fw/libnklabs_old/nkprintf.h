// Small versions of printf / sprintf

// Copyright 2020 NK Labs, LLC

// Permission is hereby granted, free of charge, to any person obtaining a
// copy of this software and associated documentation files (the
// "Software"), to deal in the Software without restriction, including
// without limitation the rights to use, copy, modify, merge, publish,
// distribute, sublicense, and/or sell copies of the Software, and to permit
// persons to whom the Software is furnished to do so, subject to the
// following conditions:

// The above copyright notice and this permission notice shall be included
// in all copies or substantial portions of the Software.

// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. 
// IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT
// OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR
// THE USE OR OTHER DEALINGS IN THE SOFTWARE.

#ifndef _nkprintf
#define _nkprintf 1

#include <stdarg.h>

#include "nkoutfile.h"
#include "nkinfile.h"
#include "nkmacros.h"

// Config options

#include "nkprintf_config.h"

// Hex dump
// Print from buf[offset] to buf[offset+count]
// Base is added to offset for the printed address

int nk_fbyte_hex_dump(nkoutfile_t *f, unsigned char *buf, unsigned long base, unsigned long offset, unsigned long count);

int nk_byte_hex_dump(unsigned char *buf, unsigned long base, unsigned long offset, unsigned long count);

// print 'ind' spaces- used for indentation

int nk_findent(nkoutfile_t *f, int ind);

int nk_indent(int ind);

// Small version of printf for small MCUs
// Low level printf
// Returns return value of emitc

// Non-standard things:
//
//  _ prints an empty string.  This is useful for variable indentation, as in nk_printf("%*_Hello, world!\n", indent);
//
//  v prints an nkdbase in serialized format, for example:   nk_printf("%v", type, location);

int nk_vprintf(nkoutfile_t *f, const char *fmt, va_list ap);

int nk_fprintf(nkoutfile_t *f, const char *fmt, ...) __attribute__((__format__ (__printf__, 2, 3)));

// print to console
// Output goes to NKPRINTF_PUTC from nkprintf_config.h

void nk_printf(const char *fmt,...) __attribute__((__format__ (__printf__, 1, 2)));

// print to string, string will always be NUL terminated
// Returns -1 if string overflowed, 0 otherwise

int nk_snprintf(char *dest, size_t len, const char *fmt,...) __attribute__((__format__ (__printf__, 3, 4)));

// print to field, field will be space filled
// Returns -1 if string overflowed, 0 otherwise

int nk_sfprintf(char *dest, size_t len, const char *fmt,...) __attribute__((__format__ (__printf__, 3, 4)));

extern nkoutfile_t *nkstdout;
extern nkoutfile_t *nkstderr;
extern nkoutfile_t *nkstdnull;

/* debug_printf */

#define nk_debug_message(fmt, ...) nk_printf(__FILE__ ":" nk_tostring(__LINE__) ": " fmt, ##__VA_ARGS__)

#define nk_error_message(fmt, ...) nk_printf(__FILE__ ":" nk_tostring(__LINE__) ": Error: " fmt, ##__VA_ARGS__)

#define nk_startup_message(fmt, ...) nk_printf("[Initialize] " fmt, ##__VA_ARGS__)

#define nk_fatal_message(fmt, ...) do { \
	nk_printf(__FILE__ ":" nk_tostring(__LINE__) ": Fatal Error: " fmt, ##__VA_ARGS__); \
	for (;;) ; \
	} while (0);

#endif
