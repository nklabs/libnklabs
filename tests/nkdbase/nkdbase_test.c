#include <stdbool.h>
#include <string.h>

#include "nkprintf.h"
#include "nkscan.h"
#include "nkdbase.h"

// Test database
// Type definitions

struct testfwd {
    int tbool;
    int tint;
    unsigned int tuint;
    char tint8;
    unsigned char tuint8;
    short tint16;
    unsigned short tuint16;
    double tdouble;
    float tfloat;
    char tstring[20];
};

struct testbkwd {
    char tstring[20];
    float tfloat;
    double tdouble;
    unsigned short tuint16;
    short tint16;
    unsigned char tuint8;
    char tint8;
    unsigned int tuint;
    int tint;
    int tbool;
};

struct testtop {
    struct testfwd tstruct;
    struct testbkwd tarray[4];
    union len tvararray_len;
    struct testfwd tvararray[8];
    union len ttable_len;
    struct testbkwd ttable[7];
};

// Test database metadata

const struct type tyTESTSTRING = {
    .what = tSTRING,
    .size = nk_member_size(struct testfwd, tstring),
    .members = NULL,
    .subtype = NULL,
    .check = NULL
};

const struct member testfwd_members[] = {
    { "tbool", &tyBOOL, offsetof(struct testfwd, tbool) },
    { "tint", &tyINT, offsetof(struct testfwd, tint) },
    { "tuint", &tyUINT, offsetof(struct testfwd, tuint) },
    { "tint8", &tyINT8, offsetof(struct testfwd, tint8) },
    { "tuint8", &tyUINT8, offsetof(struct testfwd, tuint8) },
    { "tint16", &tyINT16, offsetof(struct testfwd, tint16) },
    { "tuint16", &tyUINT16, offsetof(struct testfwd, tuint16) },
    { "tdouble", &tyDOUBLE, offsetof(struct testfwd, tdouble) },
    { "tfloat", &tyFLOAT, offsetof(struct testfwd, tfloat) },
    { "tstring", &tyTESTSTRING, offsetof(struct testfwd, tstring) },
    { NULL, NULL, 0 }
};

const struct type tyTESTFWD = {
    .what = tSTRUCT,
    .size = sizeof(struct testfwd),
    .members = testfwd_members,
    .subtype = NULL,
    .check = NULL
};

const struct member testbkwd_members[] = {
    { "tstring", &tyTESTSTRING, offsetof(struct testbkwd, tstring) },
    { "tfloat", &tyFLOAT, offsetof(struct testbkwd, tfloat) },
    { "tdouble", &tyDOUBLE, offsetof(struct testbkwd, tdouble) },
    { "tuint16", &tyUINT16, offsetof(struct testbkwd, tuint16) },
    { "tint16", &tyINT16, offsetof(struct testbkwd, tint16) },
    { "tuint8", &tyUINT8, offsetof(struct testbkwd, tuint8) },
    { "tint8", &tyINT8, offsetof(struct testbkwd, tint8) },
    { "tuint", &tyUINT, offsetof(struct testbkwd, tuint) },
    { "tint", &tyINT, offsetof(struct testbkwd, tint) },
    { "tbool", &tyBOOL, offsetof(struct testbkwd, tbool) },
    { NULL, NULL, 0 }
};

const struct type tyTESTBKWD = {
    .what = tSTRUCT,
    .size = sizeof(struct testbkwd),
    .members = testbkwd_members,
    .subtype = NULL,
    .check = NULL
};

const struct type tyTESTARRAY = {
    .what = tARRAY,
    .size = nk_member_size(struct testtop, tarray),
    .members = NULL,
    .subtype = &tyTESTBKWD,
    .check = NULL
};

const struct type tyTESTVARARRAY = {
    .what = tVARRAY,
    .size = nk_member_size(struct testtop, tvararray),
    .members = NULL,
    .subtype = &tyTESTFWD,
    .check = NULL
};

const struct type tyTESTTABLE = {
    .what = tTABLE,
    .size = nk_member_size(struct testtop, ttable),
    .members = NULL,
    .subtype = &tyTESTBKWD,
    .check = NULL
};

const struct member testtop_members[] = {
    { "tstruct", &tyTESTFWD, offsetof(struct testtop, tstruct) },
    { "tarray", &tyTESTARRAY, offsetof(struct testtop, tarray) },
    { "tvararray", &tyTESTVARARRAY, offsetof(struct testtop, tvararray_len) },
    { "ttable", &tyTESTTABLE, offsetof(struct testtop, ttable_len) },
    { NULL, NULL, 0 }
};

const struct type tyTESTTOP = {
    .what = tSTRUCT,
    .size = sizeof(struct testtop),
    .members = testtop_members,
    .subtype = NULL,
    .check = NULL
};

// Example loaded structure

struct testtop testtop = {
    .tstruct = {
        .tbool = true, .tint = 0x7FFFFFFE, .tuint = 0xEFFFFFFE, .tint8 = 0x7E, .tuint8 = 0xEE, .tint16 = 0x7FFE, .tuint16 = 0xEFFE, .tdouble = 0.125, .tfloat = .0625, .tstring = "Hello"
    },
    .tarray = {
        { .tstring = "Hello", .tfloat = .0625, .tdouble = 0.125, .tuint16 = 0xEFFE, .tint16 = 0x7FFE, .tuint8 = 0xEE, .tint8 = 0x7E, .tuint = 0xEFFFFFFE, .tint = 0x7FFFFFFE, .tbool = false },
        { .tstring = "Hello", .tfloat = .0626, .tdouble = 0.125, .tuint16 = 0xEFFE, .tint16 = 0x7FFE, .tuint8 = 0xEE, .tint8 = 0x7E, .tuint = 0xEFFFFFFE, .tint = 0x7FFFFFFE, .tbool = false },
        { .tstring = "Hello", .tfloat = .0627, .tdouble = 0.125, .tuint16 = 0xEFFE, .tint16 = 0x7FFE, .tuint8 = 0xEE, .tint8 = 0x7E, .tuint = 0xEFFFFFFE, .tint = 0x7FFFFFFE, .tbool = false },
        { .tstring = "Hello", .tfloat = .0628, .tdouble = 0.125, .tuint16 = 0xEFFE, .tint16 = 0x7FFE, .tuint8 = 0xEE, .tint8 = 0x7E, .tuint = 0xEFFFFFFE, .tint = 0x7FFFFFFE, .tbool = false },

    },
    .tvararray_len.len = 3,
    .tvararray = {
        { .tbool = true, .tint = 0x7FFFFFFE, .tuint = 0xEFFFFFFE, .tint8 = 0x7E, .tuint8 = 0xEE, .tint16 = 0x7FFE, .tuint16 = 0xEFFE, .tdouble = 0.126, .tfloat = .0625, .tstring = "Hello" },
        { .tbool = true, .tint = 0x7FFFFFFE, .tuint = 0xEFFFFFFE, .tint8 = 0x7E, .tuint8 = 0xEE, .tint16 = 0x7FFE, .tuint16 = 0xEFFE, .tdouble = 0.127, .tfloat = .0625, .tstring = "Hello" },
        { .tbool = true, .tint = 0x7FFFFFFE, .tuint = 0xEFFFFFFE, .tint8 = 0x7E, .tuint8 = 0xEE, .tint16 = 0x7FFE, .tuint16 = 0xEFFE, .tdouble = 0.128, .tfloat = .0625, .tstring = "Hello" }
    },
    .ttable_len.len = 3,
    .ttable = {
        { .tstring = "Hello", .tfloat = .0630, .tdouble = 0.125, .tuint16 = 0xEFFE, .tint16 = 0x7FFE, .tuint8 = 0xEE, .tint8 = 0x7E, .tuint = 0xEFFFFFFE, .tint = 0x7FFFFFFE, .tbool = false },
        { .tstring = "Hello", .tfloat = .0631, .tdouble = 0.125, .tuint16 = 0xEFFE, .tint16 = 0x7FFE, .tuint8 = 0xEE, .tint8 = 0x7E, .tuint = 0xEFFFFFFE, .tint = 0x7FFFFFFE, .tbool = false },
        { .tstring = "Hello", .tfloat = .0632, .tdouble = 0.125, .tuint16 = 0xEFFE, .tint16 = 0x7FFE, .tuint8 = 0xEE, .tint8 = 0x7E, .tuint = 0xEFFFFFFE, .tint = 0x7FFFFFFE, .tbool = false }
    }
};

struct testtop tryit;

int main(int argc, char *argv[])
{
    // Serialized format
    nk_dbase_serialize(nkstdout, &tyTESTTOP, &testtop);
    nk_printf("\n");

    // Printable format
    nk_dbase_fprint(nkstdout, &tyTESTTOP, &testtop, 0, "\n");

    char *ser = "{tstruct:{tbool:true, tint:2147483646, tuint:4026531838, tint8:126, tuint8:238, tint16:32766, tuint16:61438, tdouble:0.125, tfloat:0.0625, tstring:\"Hello\"}, tarray:[{tstring:\"Hello\", tfloat:0.0625, tdouble:0.125, tuint16:61438, tint16:32766, tuint8:238, tint8:126, tuint:4026531838, tint:2147483646, tbool:false}, {tstring:\"Hello\", tfloat:0.0626, tdouble:0.125, tuint16:61438, tint16:32766, tuint8:238, tint8:126, tuint:4026531838, tint:2147483646, tbool:false}, {tstring:\"Hello\", tfloat:0.0627, tdouble:0.125, tuint16:61438, tint16:32766, tuint8:238, tint8:126, tuint:4026531838, tint:2147483646, tbool:false}, {tstring:\"Hello\", tfloat:0.0628, tdouble:0.125, tuint16:61438, tint16:32766, tuint8:238, tint8:126, tuint:4026531838, tint:2147483646, tbool:false}], tvararray:[{tbool:true, tint:2147483646, tuint:4026531838, tint8:126, tuint8:238, tint16:32766, tuint16:61438, tdouble:0.126, tfloat:0.0625, tstring:\"Hello\"}, {tbool:true, tint:2147483646, tuint:4026531838, tint8:126, tuint8:238, tint16:32766, tuint16:61438, tdouble:0.127, tfloat:0.0625, tstring:\"Hello\"}, {tbool:true, tint:2147483646, tuint:4026531838, tint8:126, tuint8:238, tint16:32766, tuint16:61438, tdouble:0.128, tfloat:0.0625, tstring:\"Hello\"}], ttable:(tstring tfloat tdouble tuint16 tint16 tuint8 tint8 tuint tint tbool :\"Hello\" 0.063 0.125 61438 32766 238 126 4026531838 2147483646 false:\"Hello\" 0.0631 0.125 61438 32766 238 126 4026531838 2147483646 false:\"Hello\" 0.0632 0.125 61438 32766 238 126 4026531838 2147483646 false)}";
    nkinfile_t f[1];
    nkinfile_open_string(f, ser);
    int sta;
    sta = nk_fscan(f, "%v", &tyTESTTOP, &tryit);
    nk_printf("Parse status = %d\n", sta);
    nk_dbase_fprint(nkstdout, &tyTESTTOP, &tryit, 0, "\n");
    if (memcmp(&tryit, &testtop, sizeof(struct testtop)))
        printf("Mismatch!\n");
    else
        printf("They match!\n");
}
