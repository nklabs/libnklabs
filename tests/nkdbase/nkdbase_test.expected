{tstruct:{tbool:true, tint:2147483646, tuint:4026531838, tint8:126, tuint8:238, tint16:32766, tuint16:61438, tdouble:0.125, tfloat:0.0625, tstring:"Hello"}, tarray:[{tstring:"Hello", tfloat:0.0625, tdouble:0.125, tuint16:61438, tint16:32766, tuint8:238, tint8:126, tuint:4026531838, tint:2147483646, tbool:false}, {tstring:"Hello", tfloat:0.0626, tdouble:0.125, tuint16:61438, tint16:32766, tuint8:238, tint8:126, tuint:4026531838, tint:2147483646, tbool:false}, {tstring:"Hello", tfloat:0.0627, tdouble:0.125, tuint16:61438, tint16:32766, tuint8:238, tint8:126, tuint:4026531838, tint:2147483646, tbool:false}, {tstring:"Hello", tfloat:0.0628, tdouble:0.125, tuint16:61438, tint16:32766, tuint8:238, tint8:126, tuint:4026531838, tint:2147483646, tbool:false}], tvararray:[{tbool:true, tint:2147483646, tuint:4026531838, tint8:126, tuint8:238, tint16:32766, tuint16:61438, tdouble:0.126, tfloat:0.0625, tstring:"Hello"}, {tbool:true, tint:2147483646, tuint:4026531838, tint8:126, tuint8:238, tint16:32766, tuint16:61438, tdouble:0.127, tfloat:0.0625, tstring:"Hello"}, {tbool:true, tint:2147483646, tuint:4026531838, tint8:126, tuint8:238, tint16:32766, tuint16:61438, tdouble:0.128, tfloat:0.0625, tstring:"Hello"}], ttable:(tstring tfloat tdouble tuint16 tint16 tuint8 tint8 tuint tint tbool :"Hello" 0.063 0.125 61438 32766 238 126 4026531838 2147483646 false:"Hello" 0.0631 0.125 61438 32766 238 126 4026531838 2147483646 false:"Hello" 0.0632 0.125 61438 32766 238 126 4026531838 2147483646 false)}
{
  tstruct: {
    tbool: true,
    tint: 2147483646,
    tuint: 4026531838,
    tint8: 126,
    tuint8: 238,
    tint16: 32766,
    tuint16: 61438,
    tdouble: 0.125,
    tfloat: 0.0625,
    tstring: "Hello"
  },
  tarray: [
    {
      tstring: "Hello",
      tfloat: 0.0625,
      tdouble: 0.125,
      tuint16: 61438,
      tint16: 32766,
      tuint8: 238,
      tint8: 126,
      tuint: 4026531838,
      tint: 2147483646,
      tbool: false
    },
    {
      tstring: "Hello",
      tfloat: 0.0626,
      tdouble: 0.125,
      tuint16: 61438,
      tint16: 32766,
      tuint8: 238,
      tint8: 126,
      tuint: 4026531838,
      tint: 2147483646,
      tbool: false
    },
    {
      tstring: "Hello",
      tfloat: 0.0627,
      tdouble: 0.125,
      tuint16: 61438,
      tint16: 32766,
      tuint8: 238,
      tint8: 126,
      tuint: 4026531838,
      tint: 2147483646,
      tbool: false
    },
    {
      tstring: "Hello",
      tfloat: 0.0628,
      tdouble: 0.125,
      tuint16: 61438,
      tint16: 32766,
      tuint8: 238,
      tint8: 126,
      tuint: 4026531838,
      tint: 2147483646,
      tbool: false
    }
  ],
  tvararray: [
    {
      tbool: true,
      tint: 2147483646,
      tuint: 4026531838,
      tint8: 126,
      tuint8: 238,
      tint16: 32766,
      tuint16: 61438,
      tdouble: 0.126,
      tfloat: 0.0625,
      tstring: "Hello"
    },
    {
      tbool: true,
      tint: 2147483646,
      tuint: 4026531838,
      tint8: 126,
      tuint8: 238,
      tint16: 32766,
      tuint16: 61438,
      tdouble: 0.127,
      tfloat: 0.0625,
      tstring: "Hello"
    },
    {
      tbool: true,
      tint: 2147483646,
      tuint: 4026531838,
      tint8: 126,
      tuint8: 238,
      tint16: 32766,
      tuint16: 61438,
      tdouble: 0.128,
      tfloat: 0.0625,
      tstring: "Hello"
    }
  ],
  ttable:
    (	tstring	tfloat	tdouble	tuint16	tint16	tuint8	tint8	tuint	tint	tbool
    :	"Hello"	0.063	0.125	61438	32766	238	126	4026531838	2147483646	false
    :	"Hello"	0.0631	0.125	61438	32766	238	126	4026531838	2147483646	false
    :	"Hello"	0.0632	0.125	61438	32766	238	126	4026531838	2147483646	false
    )
}
Parse status = 1
{
  tstruct: {
    tbool: true,
    tint: 2147483646,
    tuint: 4026531838,
    tint8: 126,
    tuint8: 238,
    tint16: 32766,
    tuint16: 61438,
    tdouble: 0.125,
    tfloat: 0.0625,
    tstring: "Hello"
  },
  tarray: [
    {
      tstring: "Hello",
      tfloat: 0.0625,
      tdouble: 0.125,
      tuint16: 61438,
      tint16: 32766,
      tuint8: 238,
      tint8: 126,
      tuint: 4026531838,
      tint: 2147483646,
      tbool: false
    },
    {
      tstring: "Hello",
      tfloat: 0.0626,
      tdouble: 0.125,
      tuint16: 61438,
      tint16: 32766,
      tuint8: 238,
      tint8: 126,
      tuint: 4026531838,
      tint: 2147483646,
      tbool: false
    },
    {
      tstring: "Hello",
      tfloat: 0.0627,
      tdouble: 0.125,
      tuint16: 61438,
      tint16: 32766,
      tuint8: 238,
      tint8: 126,
      tuint: 4026531838,
      tint: 2147483646,
      tbool: false
    },
    {
      tstring: "Hello",
      tfloat: 0.0628,
      tdouble: 0.125,
      tuint16: 61438,
      tint16: 32766,
      tuint8: 238,
      tint8: 126,
      tuint: 4026531838,
      tint: 2147483646,
      tbool: false
    }
  ],
  tvararray: [
    {
      tbool: true,
      tint: 2147483646,
      tuint: 4026531838,
      tint8: 126,
      tuint8: 238,
      tint16: 32766,
      tuint16: 61438,
      tdouble: 0.126,
      tfloat: 0.0625,
      tstring: "Hello"
    },
    {
      tbool: true,
      tint: 2147483646,
      tuint: 4026531838,
      tint8: 126,
      tuint8: 238,
      tint16: 32766,
      tuint16: 61438,
      tdouble: 0.127,
      tfloat: 0.0625,
      tstring: "Hello"
    },
    {
      tbool: true,
      tint: 2147483646,
      tuint: 4026531838,
      tint8: 126,
      tuint8: 238,
      tint16: 32766,
      tuint16: 61438,
      tdouble: 0.128,
      tfloat: 0.0625,
      tstring: "Hello"
    }
  ],
  ttable:
    (	tstring	tfloat	tdouble	tuint16	tint16	tuint8	tint8	tuint	tint	tbool
    :	"Hello"	0.063	0.125	61438	32766	238	126	4026531838	2147483646	false
    :	"Hello"	0.0631	0.125	61438	32766	238	126	4026531838	2147483646	false
    :	"Hello"	0.0632	0.125	61438	32766	238	126	4026531838	2147483646	false
    )
}
They match!
